File: Python/0022.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(current, open_count, close_count):
            if open_count == n and close_count == n:
                result.append(current)
                return
            if open_count < n:
                backtrack(current + '(', open_count + 1, close_count)
            if close_count < open_count:
                backtrack(current + ')', open_count, close_count + 1)
        
        result = []
        backtrack('', 0, 0)
        return result

File: Scala/0022.scala
object Solution {
    def generateParenthesis(n: Int): List[String] = {
        // Helper function to perform recursive backtracking
        def backtrack(current: StringBuilder, pos: Int, open: Int, close: Int, n: Int, result: List[String]): List[String] = {
            if (pos == 2 * n) {
                result :+ current.toString() // Add the current combination to the result list
            } else {
                var newResult = result
                if (open < n) {
                    current.append('(')
                    newResult = backtrack(current, pos + 1, open + 1, close, n, newResult)
                    current.deleteCharAt(current.length - 1) // Remove the last character
                }
                
                if (close < open) {
                    current.append(')')
                    newResult = backtrack(current, pos + 1, open, close + 1, n, newResult)
                    current.deleteCharAt(current.length - 1) // Remove the last character
                }
                newResult
            }
        }
        
        // Initialize the backtracking process
        backtrack(new StringBuilder(), 0, 0, 0, n, List())
    }
}
